File: services/calc-engine/services/walkscore_service.py
"""
Walk Score API service for the calc engine.
Coordinates are obtained from Mapbox geocoding (via the Fastify API).
"""

import os
import httpx


async def get_walk_score(address: str, lat: float, lng: float) -> dict[str, object] | None:
    """
    Fetch Walk Score, Transit Score, and Bike Score for a property.

    Args:
        address: Full property address.
        lat: Property latitude.
        lng: Property longitude.

    Returns:
        Dict with walk_score, transit_score, bike_score, description.
        Returns None if the API call fails.
    """
    api_key = os.environ.get('WALKSCORE_API_KEY')
    if not api_key:
        return None

    url = (
        f"https://api.walkscore.com/score"
        f"?format=json"
        f"&address={httpx.URL(address)}"
        f"&lat={lat}"
        f"&lon={lng}"
        f"&transit=1"
        f"&bike=1"
        f"&wsapikey={api_key}"
    )

    try:
        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.get(url)
            response.raise_for_status()
            data = response.json()
            return {
                'walk_score': data.get('walkscore', 0),
                'transit_score': data.get('transit', {}).get('score', 0),
                'bike_score': data.get('bike', {}).get('score', 0),
                'description': data.get('description', ''),
            }
    except Exception:
        return None

File: apps/api/src/services/walkScoreService.ts
/* eslint-disable no-console */
/**
 * Walk Score + Transit Score API service.
 * Docs: https://www.walkscore.com/professional/api.php
 */

import type { WalkScoreResult } from '../types/analysis'

export type { WalkScoreResult }

/**
 * Fetch Walk Score, Transit Score, and Bike Score for a given address.
 * Returns null on any failure — the orchestrator degrades gracefully without scores.
 */
export async function getWalkScore(
  address: string,
  lat: number,
  lng: number
): Promise<WalkScoreResult | null> {
  try {
    const WALKSCORE_API_KEY = process.env.WALKSCORE_API_KEY ?? ''

    if (!WALKSCORE_API_KEY) {
      console.warn('getWalkScore: WALKSCORE_API_KEY is not set — skipping walk score')
      return null
    }

    const url =
      `https://api.walkscore.com/score` +
      `?format=json` +
      `&address=${encodeURIComponent(address)}` +
      `&lat=${lat}` +
      `&lon=${lng}` + // Walk Score uses 'lon', not 'lng'
      `&transit=1` +
      `&bike=1` +
      `&wsapikey=${WALKSCORE_API_KEY}`

    const controller = new AbortController()
    const timeout = setTimeout(() => controller.abort(), 10_000)

    let res: Response
    try {
      res = await fetch(url, { signal: controller.signal })
    } finally {
      clearTimeout(timeout)
    }

    if (!res.ok) {
      console.warn(`getWalkScore: Walk Score API returned ${res.status} for "${address}"`)
      return null
    }

    const data = (await res.json()) as {
      walkscore?: number
      description?: string
      transit?: { score?: number }
      bike?: { score?: number }
    }

    if (data.walkscore == null) {
      return null
    }

    return {
      walk: data.walkscore,
      transit: data.transit?.score ?? null,
      bike: data.bike?.score ?? null,
      description: data.description ?? '',
    }
  } catch (err) {
    console.error(`getWalkScore: error fetching walk score for "${address}":`, err)
    return null
  }
}
